File: LC/[1092] Shortest Common Supersequence.py
class Solution:
    def shortestCommonSupersequence(self, str1: str, str2: str) -> str:
        """
        N=
        => shortest string such that both str1, str2 are its subsequence
        """
        s1 = '#'+str1
        s2 = '#'+str2
        N1 = len(s1)
        N2 = len(s2)
        dp = [[math.inf]*N2 for _ in range(N1)]

        # initial states
        for ii in range(N1):
            dp[ii][0] = ii
        for ii in range(N2):
            dp[0][ii] = ii

        for ii in range(1, N1):
            for jj in range(1, N2):
                if s1[ii] == s2[jj]:
                    dp[ii][jj] = dp[ii-1][jj-1]+1
                else:
                    """
                    x x x x ii 
                    y y y y y y jj x
                    """
                    add_s1 = dp[ii-1][jj]
                    add_s2 = dp[ii][jj-1]

                    dp[ii][jj] = 1+min(add_s1, add_s2)

        # reconstruct the path
        res = []
        ii, jj = N1-1, N2-1
        while ii > 0 and jj > 0:
            """
            x x x x x a
            y y y b
            """
            # if dp[ii][jj]==(dp[ii-1][jj-1]+1): << why this is not working?
            # s1[ii]==s2[jj] means dp[ii][jj]==dp[ii-1][jj-1], but not the other way around
            if s1[ii] == s2[jj]:
                res.append(s1[ii])
                ii -= 1
                jj -= 1
            elif dp[ii][jj] == (dp[ii-1][jj]+1):
                res.append(s1[ii])
                ii -= 1
            else:
                res.append(s2[jj])
                jj -= 1

        while ii > 0:
            res.append(s1[ii])
            ii -= 1
        while jj > 0:
            res.append(s2[jj])
            jj -= 1

        # print("res: ", "".join(res))
        return "".join(res[::-1])

File: LC/[1092] Shortest Common Supersequence.java
class Solution {
    public String shortestCommonSupersequence(String str1, String str2) {
        int N1 = str1.length();
        int N2 = str2.length();

        String s = "#" + str1; //since we need previous state for index 0
        String t = "#" + str2;

        //declare dp
        int[][] dp = new int[N1+1][N2+1]; //consider the dummy header
        /*
        dp[ii][jj]: min len of a ss s.t. s and t are both subsequence of ss
        */

        //initialize dp
        //when ii=0
        for(int jj=1; jj<=N2; ++jj) dp[0][jj] = jj;
        //when jj=0
        for(int ii=0; ii<=N1; ++ii) dp[ii][0] = ii;
        dp[0][0]=0;
        //populate dp
        for(int ii=1; ii<=N1; ++ii){//from 1 due to dummy header
            for(int jj=1; jj<=N2; ++jj){
                if(s.charAt(ii)==t.charAt(jj)){
                    dp[ii][jj] = dp[ii-1][jj-1] +1;
                }else{
                    dp[ii][jj] = Math.min(dp[ii-1][jj], dp[ii][jj-1]) + 1;
                }
            }
        }

        String ret="";
        //backtrack for the path
        int ii = N1;
        int jj = N2;

        while(ii>0 && jj>0){
            if(s.charAt(ii)==t.charAt(jj)){
                ret = s.charAt(ii)+ret;
                --ii;
                --jj;
            }else{
                /*
                xxxxxa
                yyyyb
                zzzzz a
                */
                if(dp[ii][jj]==dp[ii-1][jj]+1){
                    ret = s.charAt(ii--)+ret;
                }else{
                    ret = t.charAt(jj--)+ret;
                }
            }
        }

        while(ii > 0) ret = s.charAt(ii--)+ret; 
        while(jj > 0) ret = t.charAt(jj--)+ret; 
            
        return ret;        
    }
}

/*
Requirement:
* Given s, t two strings, find shortest string ss s.t. s and t are both ss's subsequence
Analysis:
* N: 1000 -> two loops
* two series looking for optimized metrics -> DP III

*/